### PR TITLE
update logging validate

### PR DESCRIPTION
**Pull request by aiwantaozi**
@zionwu @gitlawr 
this PR is for update endpoint validate, I just miss the validate in kafka zookeeper, not add it. And the replace url.parse with url.parserequesturi, could u you help review.

### CODE REVIEW

File: pkg/controllers/user/logging/utils/wrap.go [hunk 1]
```diff
@@ -159,17 +160,52 @@ func getWrapConfig(es *v3.ElasticsearchConfig, sp *v3.SplunkConfig, sl *v3.Syslo
...
+		}
+		currentTarget = loggingconfig.Kafka
 	}
 	return
 }
 
+func parseEndpoint(endpoint string) (host string, scheme string, err error) {
+	u, err := url.ParseRequestURI(endpoint)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "invaild endpoint %s", endpoint)
```
Comment (gitlawr) [inline pkg/controllers/user/logging/utils/wrap.go]: typo `invaild`, same as the below one.

**Comment by zionwu** [comment]
LGTM

**Comment by zionwu** [comment]
https://github.com/rancher/rancher/issues/11736

File: pkg/api/customization/logging/list_handler.go [hunk 2]
```diff
@@ -153,7 +153,7 @@ func (h *ClusterLoggingHandler) setEmbeddedEndpoint(cl *v3.ClusterLogging) error
 	} else {
 		kibanapod := kibanapods[0]
 		if kibanapod.Status.Phase == running {
-			cl.Spec.EmbeddedConfig.KibanaEndpoint = fmt.Sprintf("%s:%v", kibanapod.Status.HostIP, esPort)
+			cl.Spec.EmbeddedConfig.KibanaEndpoint = fmt.Sprintf("http://%s:%v", kibanapod.Status.HostIP, kibanaservice.Spec.Ports[0].NodePort)
```
Comment (alena1108) [inline pkg/api/customization/logging/list_handler.go]: any chance ports len can be == 0?

File: pkg/controllers/user/logging/utils/wrap.go [hunk 3]
```diff
@@ -159,17 +160,52 @@ func getWrapConfig(es *v3.ElasticsearchConfig, sp *v3.SplunkConfig, sl *v3.Syslo
...
+			err = errors.New("one of the kafka endpoint must be set")
+			return
 		}
+		if len(kf.BrokerEndpoints) != 0 {
+			var bs []string
+			var h string
+			for _, v := range kf.BrokerEndpoints {
+				h, _, err = parseEndpoint(v)
+				if err != nil {
+					return
```
Comment (alena1108) [inline pkg/controllers/user/logging/utils/wrap.go]: should we log the error?

File: pkg/controllers/user/logging/utils/wrap.go [hunk 4]
```diff
@@ -159,17 +160,52 @@ func getWrapConfig(es *v3.ElasticsearchConfig, sp *v3.SplunkConfig, sl *v3.Syslo
...
+				}
+				bs = append(bs, h)
+			}
+			wkf = WrapKafka{
+				Brokers: strings.Join(bs, ","),
+			}
+		} else {
+			if kf.ZookeeperEndpoint != "" {
+				if _, _, err = parseEndpoint(kf.ZookeeperEndpoint); err != nil {
+					return
```
Comment (alena1108) [inline pkg/controllers/user/logging/utils/wrap.go]: same here - is it worth logging the error?

File: pkg/api/customization/logging/list_handler.go (hunk 2 above)
Comment (aiwantaozi) [reply pkg/api/customization/logging/list_handler.go]: if the length is 0, it return error in the line above(line 147)

File: pkg/controllers/user/logging/utils/wrap.go (hunk 4 above)
Comment (aiwantaozi) [reply pkg/controllers/user/logging/utils/wrap.go]: it will return the message to UI, when the validate is failed, may be don't need to log the error.

**Comment by alena1108** [comment]
LGTM

**Comment by zionwu** [comment]
https://github.com/rancher/rancher/issues/11818

**Pull request merged**